### ISSUE TITLE
Notation formats are not associated to the right notation

### ISSUE THREAD

**Issue by ppedrot** [coq/coq#6082]
The following code displays the wrong behaviour.

```
Notation "[ x ]" := (S x) (format "[  x  ]") : foo.
Notation "[ x ]" := (S x) (format "[ x ]") : bar.

Open Scope bar.

Check fun x => S x.
(** fun x : nat => [ x ] *)
```
The correct printing should rather be `fun x : nat => [x]`, without additional spacing.

This bug is due to the fact that formats are globally shared by all notations having the same key, i.e. their shape, which is `[ _ ]` in this case.

Adding @herbelin to the discussion as he's probably the most informed one on the question.

**Comment by herbelin** [comment]
Indeed, the currrent bug is that (some) modifiers are silently ignored in general (see #5740) while it should be an error since there is no support to change them.

The feature request is then of course to be able to have different printing rules for different scopes, since, on the contrary of a token-based parsing, it is possible to support it at printing. Don't hesitate to provide a PR as you mentioned at #6078.

**Comment by ppedrot** [comment]
> Don't hesitate to provide a PR as you mentioned at #6078.

@herbelin The problem is that I see in theory how to fix the problem with a partial rewriting of the code for notations, but this is going to break the ML API between 8.7.1 and 8.7.2, so this doesn't look good... The fix for #6078 in 8.7 should be small and self-contained, even it that pushes it on the fringes of hackery.

**Comment by herbelin** [comment]
For #6078, I suspect the simpler fix is to remove the ssreflect notation. As far as I can judge, the comment expresses that the notation is supposed to do what the primitive printer is [exactly](https://github.com/coq/coq/blob/master/printing/ppconstr.ml#L709-L718) doing. So, I don't see what it can bring good.

And I don't know how to support several printing rules depending on scope without changing the API, so I would a priori not care about #6078 for this improvement.

If #6078 is your priority, wouldn't it be enough to overwrite the printing rule (independently of scopes) by warning rather than failing?
